Ikuyadeu/vscode-R
# Use fs.watch instead of vscode.FileSystemWatcher

* Close #347
* Close #352 
* Support multi-root workspace (Close #236)
* Support workspace in cloud drive (Close #179, Close #272, Close #330) and read-only file systems

This PR uses `fs.watch` to replace `vscode.FileSystemWatcher` and the `response.log` renamed to `request.log` and is moved to `~/.vscode-R/request.log`; `globalenv.json` and `plot.png` are moved to `{tempdir}/vscode-R/`.

Note that `fs.watch` may trigger multiple events when the watched file is being written. If we directly watch `request.log`, `globalenv.json`, and `plot.png`, respectively, it is likely that we might read a partial file that is still being written. To avoid this, I use lock files for each of them, i.e., `resquest.lock`, `globalenv.lock`, `plot.lock`. In `init.R`, these lock files are written a timestamp after the underlying file is finished writing. Only when the file watch triggers an event and the timestamp changes do we read or open the underlying file to avoid duplicate reading or partial reading.

All functionalities of session watcher should not change. The only visible change is that it no longer creates `{workspaceFolder}/.vscode/vscode-R`.

**Testing this PR**

* General functionalities
  - attach
  - symbol hover
  - `View(mtcars)`
  - `View(baseenv())`
  - `plot(rnorm(100))`
  - R session started from a subfolder of the VSCode workspace root should work now.
* Multi-session support
  - Open a workspace at folder A
  - Create multiple terminals
  - Each should be able to attach
  - Click status bar in previously created terminal and it should attach
  - In the attached session, `plot(rnorm(100))` should open the png file
  - In any of the sessions, `View(mtcars)` should trigger the WebView
* Multi-root workspace support
  - Create a multi-root workspace
  - Start an R session from one workspace folder should work
  - Start an R session from the second workspace folder should work too
* Cloud-drive support
  - Create a workspace in a cloud-drive (e.g. OneDrive, Google Drive)
  - Start an R session from the workspace folder
  - The session watcher should work normally
* Read-only file system support
  - Open a workspace in a read-only file system (e.g. a read-only mounted drive, or non-writable directory for our user)
  - Start an R session from the workspace folder
  - The session watcher should work normally
